fq_test.go

```diff
@@ -0,0 +1,275 @@
+		if desc.imin == desc.imax {
+			it.size = desc.imax
+		} else {
+			it.size = desc.imin + uint64(rand.Int63n(int64(desc.imax-desc.imin)))
+		}
+		if t+it.size > desc.ftotal {
+			it.size = desc.ftotal - t
+		}
+		t += it.size
+		it.seq = i
+		// new packet
+		fq.Enqueue(it)
+	}
+	(*done_wg).Done()
+}
+
+func consumeQueue(t *testing.T, fq *FQScheduler, descs []flowDesc) (float64, error) {
+	active := make(map[uint64]bool)
+	var total uint64
+	acnt := make(map[uint64]uint64)
+	cnt := make(map[uint64]uint64)
+	seqs := make(map[uint64]uint64)
+
+	var wsum uint64
+	for range descs {
+		wsum += uint64(1)
+	}
+
+	// for !fq.seen {
+
+	// }
+	// TODO(aaron-prindle) remove this
+	time.Sleep(1 * time.Second)
+	for i, ok := fq.processround(); ok; i, ok = fq.processround() {
+		time.Sleep(time.Microsecond) // Simulate constrained bandwidth
+		it := i
+		seq := seqs[it.key]
+		if seq+1 != it.seq {
+			return 0, fmt.Errorf("Packet for flow %d came out of queue out-of-order: expected %d, got %d", it.key, seq+1, it.seq)
+		}
+		seqs[it.key] = it.seq
+
+		// set the flow this item is a part of to active
+		if cnt[it.key] == 0 {
+			active[it.key] = true
+		}
+		cnt[it.key] += it.size
+
+		// if # of active flows is equal to the # of total flows, add to total
+		// we are correctly taking a bit from each
+		if len(active) == len(descs) {
+			acnt[it.key] += it.size
+			total += it.size
+		}
+
+		// if all items have been processed from the flow, remove it from active
+		if cnt[it.key] == descs[it.key].ftotal {
+			delete(active, it.key)
+		}
+	}
+
+	var variance float64
+	for key := uint64(0); key < uint64(len(descs)); key++ {
+		if total == 0 {
+			t.Fatalf("expected 'total' to be nonzero")
+		}
+		// idealPercent = total-all-active/len(flows) / total-all-active
+		// idealPercent = ~((10000/10) / 10000) * 100 = ~10
+		// "how many bytes/requests you expect for this flow - all-active"
+		descs[key].idealPercent = (((float64(total)) / float64(wsum)) / float64(total)) * 100
+
+		// actualPercent = requests-for-this-flow-all-active / total-reqs
+		// "how many bytes/requests you got for this flow - all-active"
+		descs[key].actualPercent = (float64(acnt[key]) / float64(total)) * 100
+
+		x := descs[key].idealPercent - descs[key].actualPercent
+		x *= x
+		variance += x
+	}
+	fmt.Println(descs)
+
+	stdDev := math.Sqrt(variance)
+	return stdDev, nil
+}
+
+func TestSingleFlow(t *testing.T) {
+	runtime.GOMAXPROCS(runtime.NumCPU())
+	queues := initQueues(1, 1)
+	fq := newFQScheduler(queues)
+
+	go func() {
+		for i := 1; i < 10000; i++ {
+			it := &Packet{}
+			it.key = 1
+			it.size = uint64(rand.Int63n(10) + 1)
+			it.seq = uint64(i)
+			fq.Enqueue(it)
+		}
+	}()
+
+	var seq uint64
+	hasEntered := false
+	for hasEntered {
+		for it, ok := fq.Dequeue(); ok; it, ok = fq.Dequeue() {
+			if seq+1 != it.seq {
+				t.Fatalf("Packet came out of queue out-of-order: expected %d, got %d", seq+1, it.seq)
+			}
+			seq = it.seq
+		}
+	}
+}
+
+func TestTimingFlow(t *testing.T) {
```

Comment:
What does this test exercise that is not exercised by `TestUniformMultiFlow` ?

Reply:
Sorry, I had been hacking on things which is why some of this is in disarray.  TestUniformMultiFlow is the primary test that submits all requests and then dequeues all requests.  I thought it would also make sense to have something that had a running process of submitting requests and a running process of dequeuing to verify that case.  Currenty the tests are similar as the FQ skeleton does not have these runner/manager processses.